File: everything/src/tools/create-post.js
import { McpError, ErrorCode } from '@modelcontextprotocol/sdk/types.js';

/**
 * Handler function for creating a new post via Postiz API
 * 
 * The Postiz API expects a specific structure for creating posts:
 * {
 *   type: 'draft' | 'schedule' | 'now',
 *   shortLink: boolean,
 *   date: ISO date string,
 *   tags: [],
 *   posts: [{
 *     integration: { id: string },
 *     value: [{ content: string, image: [] }],
 *     settings: { __type: provider_identifier }
 *   }]
 * }
 */
export async function handleCreatePost(api, args) {
    try {
        // Validate API instance
        if (!api) {
            throw new McpError(
                ErrorCode.InternalError,
                'API instance not initialized'
            );
        }
        
        // Destructure and validate arguments
        const { 
            content, 
            integration_id, 
            type = 'now', 
            date, 
            media_ids = [], 
            shortLink = false 
        } = args;
        
        if (!content || typeof content !== 'string') {
            throw new McpError(
                ErrorCode.InvalidParams,
                'Content parameter is required and must be a string'
            );
        }
        
        if (!integration_id) {
            throw new McpError(
                ErrorCode.InvalidParams,
                'integration_id is required. Use get-integrations to find available integration IDs.'
            );
        }
        
        if (!date) {
            throw new McpError(
                ErrorCode.InvalidParams,
                'date is required in ISO 8601 format (e.g., 2025-01-04T19:46:00.000Z)'
            );
        }
        
        if (!['draft', 'schedule', 'now'].includes(type)) {
            throw new McpError(
                ErrorCode.InvalidParams,
                'type must be one of: "draft", "schedule", "now"'
            );
        }

        // First, get the integration details to find the provider identifier
        let providerIdentifier;
        try {
            const integrationsResponse = await api.get('/public/v1/integrations');
            const integrations = integrationsResponse.data;
            const integration = integrations.find(i => i.id === integration_id);
            
            if (!integration) {
                throw new McpError(
                    ErrorCode.InvalidParams,
                    `Integration with ID ${integration_id} not found. Use get-integrations to see available integrations.`
                );
            }
            
            providerIdentifier = integration.identifier;
        } catch (error) {
            if (error instanceof McpError) throw error;
            throw new McpError(
                ErrorCode.InternalError,
                `Failed to fetch integration details: ${error.message}`
            );
        }

        // Build the media/image array from media_ids
        const imageArray = media_ids.map(id => ({ id }));

        // Build the request payload in the format expected by Postiz API
        const requestData = {
            type,
            shortLink,
            date,
            tags: [],
            posts: [{
                integration: { id: integration_id },
                value: [{
                    content,
                    image: imageArray
                }],
                settings: {
                    __type: providerIdentifier
                }
            }]
        };
        
        console.error('Creating post with payload:', JSON.stringify(requestData, null, 2));
        
        // Make API request
        const response = await api.post('/public/v1/posts', requestData);

        // Check for error response
        if (response.status >= 400) {
            throw new McpError(
                ErrorCode.InternalError,
                `API error (${response.status}): ${JSON.stringify(response.data)}`
            );
        }

        return {
            content: [{
                type: 'text',
                text: JSON.stringify({
                    success: true,
                    message: 'Post created successfully',
                    data: response.data
                }, null, 2)
            }]
        };
        
    } catch (error) {
        // Handle different error types
        if (error instanceof McpError) {
            throw error; // Re-throw MCP errors as-is
        }
        
        if (error.response) {
            // API error response
            const status = error.response.status;
            const message = error.response.data?.message || error.response.data?.msg || error.message;
            
            if (status === 401) {
                throw new McpError(
                    ErrorCode.InvalidParams,
                    'Authentication failed. Please check your API key.'
                );
            } else if (status === 403) {
                throw new McpError(
                    ErrorCode.InvalidParams,
                    'Access forbidden. Check your permissions.'
                );
            } else if (status === 429) {
                throw new McpError(
                    ErrorCode.InternalError,
                    'Rate limit exceeded. Please try again later.'
                );
            } else {
                throw new McpError(
                    ErrorCode.InternalError,
                    `API error (${status}): ${message}`
                );
            }
        }
        
        // Network or other errors
        throw new McpError(
            ErrorCode.InternalError,
            `Failed to create post: ${error.message}`
        );
    }
}

/**
 * Tool definition for create_post
 */
export const createPostToolDefinition = {
    name: 'create_post',
    description: 'Create a new social media post on Postiz platform',
    inputSchema: {
        type: 'object',
        properties: {
            content: {
                type: 'string',
                description: 'Text content of the post'
            },
            integration_id: {
                type: 'string', 
                description: 'Integration ID for the target platform (use get-integrations to find available IDs)'
            },
            type: {
                type: 'string',
                description: 'Type of post: "draft", "schedule", or "now"',
                enum: ['draft', 'schedule', 'now'],
                default: 'now'
            },
            date: {
                type: 'string',
                description: 'Publish date in ISO 8601 format (e.g., "2025-01-04T19:46:00.000Z")'
            },
            media_ids: {
                type: 'array',
                items: {
                    type: 'string'
                },
                description: 'Array of media IDs (from uploaded media)',
                default: []
            },
            shortLink: {
                type: 'boolean',
                description: 'Whether to create short links',
                default: false
            }
        },
        required: ['content', 'integration_id', 'date']
    }
};

File: everything/src/tools/create_post.py
"""
Tool to create a new post using Postizz API
"""
import requests
import json
from typing import Dict, Any, Optional, List

def create_post(
    content: str,
    integration_id: str,
    post_type: str = "now",
    publish_date: Optional[str] = None,
    media_urls: Optional[List[str]] = None,
    short_link: bool = False
) -> Dict[str, Any]:
    """
    Create a new post via Postizz API
    
    Args:
        content: Text content of the post (minimum 6 characters)
        integration_id: Integration ID to post to
        post_type: Type of post ("draft", "schedule", or "now")
        publish_date: Optional ISO format date (e.g. "2025-01-04T19:46:00.000Z")
        media_urls: Optional list of media URLs to attach
        short_link: Whether to create a short link
        
    Returns:
        Dict containing created post data
    """
    if len(content) < 6:
        raise ValueError("Content must be at least 6 characters long")
    
    if post_type not in ["draft", "schedule", "now"]:
        raise ValueError('post_type must be one of: "draft", "schedule", "now"')
    
    API_URL = "https://postiz.oculair.ca/api/public/v1/posts"
    headers = {
        "Authorization": "255364482e25a51ff9a6660135b8c88995e7ae68615fb9a67ad14d601f8fe0ff",
        "Content-Type": "application/json"
    }
    
    # Build post value array
    value = [{"content": content}]  # Using "content" field
    if media_urls:
        for url in media_urls:
            value.append({"media": url})
    
    post = {
        "integration": {  # Single object with id field
            "id": integration_id
        },
        "value": value
    }
    
    post_data = {
        "type": post_type,
        "shortLink": short_link,
        "date": publish_date or "2025-02-27T19:30:00.000Z",
        "posts": [post]
    }
    
    print(f"Making request to {API_URL}")
    print(f"Headers: {json.dumps(headers, indent=2)}")
    print(f"Post data: {json.dumps(post_data, indent=2)}")
    
    try:
        response = requests.post(API_URL, headers=headers, json=post_data)
        print(f"\nResponse Status Code: {response.status_code}")
        print(f"Response Headers: {json.dumps(dict(response.headers), indent=2)}")
        
        response.raise_for_status()
        data = response.json()
        print(f"\nResponse Data: {json.dumps(data, indent=2)}")
        return data
        
    except requests.exceptions.RequestException as e:
        error_message = f"Failed to create post: {str(e)}"
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_message += f"\nResponse Status: {e.response.status_code}"
                error_message += f"\nResponse Body: {e.response.text}"
            except:
                pass
        print(f"\nError: {error_message}")
        raise Exception(error_message)

if __name__ == "__main__":
    try:
        # Example usage
        post = create_post(
            content="Testing the Postizz API integration",
            integration_id="cm5hudkk000017412anjxa09j",
            post_type="now",
            publish_date="2025-02-27T19:30:00.000Z",
            short_link=False
        )
    except Exception as e:
        print(f"\nScript failed: {str(e)}")